File: HAI704I - Architectures Distribuées/TP2/TP2-Server/TP2-ServiceServer/OfferCache.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using TP2_ServiceServer.model.offer;

namespace TP2_ServiceServer
{
    public class OfferCache
    {
        private IList<Offer> cached;

        public OfferCache()
        {
            this.cached = new List<Offer>();
        }

        public void CacheAll(ICollection<Offer> offers)
        {
            foreach(Offer offer in offers)
            {
                this.Cache(offer);
            }
        }

        public void Cache(Offer offer)
        {
            offer.id = this.GetNextId();
            this.cached.Add(offer);
        }

        public Offer Uncache(int id)
        {
            foreach(Offer offer in this.cached)
            {
                Console.WriteLine(offer.id + " ?= " + id);
                if(offer.id == id)
                {
                    this.cached.Remove(offer);
                    return offer;
                }
            }

            return null;
        }

        private int GetNextId()
        {
            if(this.cached.Count == 0)
            {
                return 0;
            }
            else
            {
                return (from offer in this.cached select offer.id).Max() + 1;
            }
        }
    }
}

File: HAI704I - Architectures Distribuées/TP2/TP2-Java/TP2-Server/src/main/java/tp2/server/service/OfferCache.java
package tp2.server.service;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.OptionalInt;

import tp2.server.model.offer.Offer;

public class OfferCache
{
	private List<Offer> cached;

	public OfferCache()
	{
		this.cached = new ArrayList<Offer>();
	}

	public void cacheAll(Collection<Offer> offers)
	{
		for(Offer offer : offers)
		{
			this.cache(offer);
		}
	}

	public void cache(Offer offer)
	{
		offer.setId(this.getNextId());
		this.cached.add(offer);
	}

	public Offer uncache(int id)
	{
		for(Offer offer : this.cached)
		{
			if(offer.getId() == id)
			{
				this.cached.remove(offer);
				return offer;
			}
		}

		return null;
	}

	private int getNextId()
	{
		OptionalInt max = this.cached.stream().mapToInt(offer -> {
			return offer.getId();
		}).max();

		if(max.isPresent())
		{
			return max.getAsInt() + 1;
		}
		else
		{
			return 0;
		}
	}
}